### PR TITLE
[C] Clear OpenTK namespace

### LINKED ISSUE #13160: .NET: Clear the OpenTK namespace
# Types in the OpenTK namespace

We want to remove types from the OpenTK namespace, and these types fall in three categories:

1. Types we can replace with an existing .NET type
2. Types that don't have an equivalent .NET type, so we have to come up with an alternative type.
3. Types we don't need, so we can remove them.

One important question is which namespace to use for 2. Options:

1. Keep using `OpenTK`. This is not an option, because that's exactly what we're trying to move away from.
2. `ObjCRuntime`: not really very math-like, and this namespace shouldn't become a catch-all for everything that doesn't fit elsewhere either.
3. `CoreNumerics`: A mix of Apple's Core* frameworks with System.Numerics.
4. `CoreMath`: same as 3.
5. `CoreGraphics`: existing Apple namespace. Con: might end up with a name clash in the future depending on what Apple does.
5. Something else?

Ref: https://github.com/xamarin/xamarin-macios/pull/2571
Ref: https://bugzilla.xamarin.com/show_bug.cgi?id=58599

## Different type

The following types have an equivalent in .NET, so the proposal is to use that type instead:

| OpenTK type       | .NET type                            |
| ----------------- | ------------------------------------ |
| OpenTK.Matrix4    | System.Numerics.Matrix4x4            |
| OpenTK.Quaternion | System.Numerics.Quaternion           |
| OpenTK.Vector2    | System.Numerics.Vector2              |
| OpenTK.Vector2d   | System.Numerics.Vector&lt;double&gt; |
| OpenTK.Vector2i   | System.Numerics.Vector&lt;int&gt;    |
| OpenTK.Vector3    | System.Numerics.Vector3              |
| OpenTK.Vector3d   | System.Numerics.Vector&lt;double&gt; |
| OpenTK.Vector3i   | System.Numerics.Vector&lt;int&gt;    |
| OpenTK.Vector4    | System.Numerics.Vector4              |
| OpenTK.Vector4d   | System.Numerics.Vector&lt;double&gt; |
| OpenTK.Vector4i   | System.Numerics.Vector&lt;int&gt;    |

## Different namespace

The following OpenTK types have no equivalent in .NET, so the proposal is to copy the OpenTK implementation, but use a different namespace.

* OpenTK.Matrix2
* OpenTK.Matrix3
* OpenTK.Matrix4d
* OpenTK.Quaterniond

The following are types that we created ourselves, but put in the OpenTK namespace.

* OpenTK.NMatrix4d
* OpenTK.NMatrix2
* OpenTK.NMatrix3
* OpenTK.NMatrix4x3
* OpenTK.NMatrix4
* OpenTK.NVector3d
* OpenTK.NVector3

## Types to remove

The following types don't seem to be used by any other API, so they can be removed

* OpenTK.Box2
* OpenTK.Functions
* OpenTK.Half
* OpenTK.Vector2h
* OpenTK.Vector3h
* OpenTK.Vector4h
* OpenTK.BezierCurve
* OpenTK.BezierCurveCubic
* OpenTK.BezierCurveQuadric

## Don't know yet

* ❓OpenTK.MathHelper

Ref: https://github.com/xamarin/xamarin-macios/issues/13087

### PR DESCRIPTION
**Pull request by StephaneDelcroix**
fixes #13160

- remove unused types
- use System.Numerics when possible
- move own created types from OpenTK namespace to CoreGraphics
- create missing types in CoreGraphics namespace

### CODE REVIEW

File: src/OpenGL/OpenTK/Math/Quaternion.cs [hunk 1]
```diff
@@ -67,21 +68,25 @@ public Quaternion(float x, float y, float z, float w)
             : this(new Vector3(x, y, z), w)
         { }
 
+#if NET
```
Comment (rolfbjarne) [inline src/OpenGL/OpenTK/Math/Quaternion.cs:71]: The whole file is inside an `!NET`, so these changes shouldn't be needed.

File: src/OpenGL/OpenTK/Math/Vector2.cs [hunk 2]
```diff
@@ -897,6 +898,7 @@ public static void BaryCentric(ref Vector2 a, ref Vector2 b, ref Vector2 c, floa
 
         #region Transform
 
+#if !NET
```
Comment (rolfbjarne) [inline src/OpenGL/OpenTK/Math/Vector2.cs:901]: There's already a file-scoped `#if !NET`

File: src/OpenGL/OpenTK/Math/Vector3.cs [hunk 3]
```diff
@@ -83,6 +84,7 @@ public Vector3(float x, float y, float z)
             Z = z;
         }
 
+#if !NET
```
Comment (rolfbjarne) [inline src/OpenGL/OpenTK/Math/Vector3.cs:87]: Already file-based `#if !NET` (and the same later in this file as well)

File: src/OpenGL/OpenTK/Math/Vector4.cs [hunk 4]
```diff
@@ -124,6 +125,7 @@ public Vector4(float x, float y, float z, float w)
             W = w;
         }
 
+#if !NET
```
Comment (rolfbjarne) [inline src/OpenGL/OpenTK/Math/Vector4.cs:128]: Same

File: src/SceneKit/SCNVector4.cs [hunk 5]
```diff
@@ -1103,6 +1126,46 @@ public bool Equals(SCNVector4 other)
...
+        /// </summary>
+        /// <param name="x">A number.</param>
+        /// <returns>An approximation of the inverse square root of the specified number, with an upper error bound of 0.001</returns>
+        /// <remarks>
+        /// This is an improved implementation of the the method known as Carmack's inverse square root
+        /// which is found in the Quake III source code. This implementation comes from
+        /// http://www.codemaestro.com/reviews/review00000105.html. For the history of this method, see
+        /// http://www.beyond3d.com/content/articles/8/
+        /// </remarks>
+        static float InverseSqrtFast(float x)
```
Comment (rolfbjarne) [inline src/SceneKit/SCNVector4.cs:1141]: This is the second time InverseSqrtFast shows up, maybe make one of the `internal static` and call it elsewhere?

File: src/Simd/QuaternionDouble.cs [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+		{
+			x = xyz.X;
+			y = xyz.Y;
+			z = xyz.Z;
+			this.w = w;
+		}
+
+		public NVector3d Xyz
+		{
+		    get => new NVector3d(x, y, z);
```
Comment (rolfbjarne) [inline src/Simd/QuaternionDouble.cs:44]:
```suggestion
		    get => new NVector3d (x, y, z);
```

File: src/Simd/QuaternionDouble.cs [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+		internal void Normalize()
+		{
+			double scale = 1.0f / this.Length;
+			Xyz *= scale;
+			W *= scale;
+		}
+
+		internal void ToAxisAngle(out NVector3d axis, out double angle)
+		{
+		    NVector4d result = ToAxisAngle();
```
Comment (rolfbjarne) [inline src/Simd/QuaternionDouble.cs:127]: Code formatting isn't following our style guide, the entire file should be re-formatted.

File: src/glkit.cs [hunk 8]
```diff
@@ -247,6 +251,7 @@ interface GLKEffectPropertyTexture {
 
 	}
 
+#if !NET
```
Comment (rolfbjarne) [inline src/glkit.cs:254]: What's the problem with `GLKEffectPropertyTransform`?

File: src/spritekit.cs [hunk 9]
```diff
@@ -1856,7 +1867,7 @@ interface SKUniform : NSCopying, NSSecureCoding {
 #else
 		[Internal]
 		IntPtr InitWithNameMatrixFloat2x2 (string name, Matrix2 value);
-#endif
+#endif // WATCH
 #endif // !XAMCORE_4_0
```
Comment (rolfbjarne) [inline src/spritekit.cs:1871]:
```suggestion
#endif // !NET
```

**Review by rolfbjarne — CHANGES_REQUESTED** [review]
First pass done! Looks good, although I'd like to know the reason for the removals related to Matrix3/Matrix4.

Also: it doesn't look like the opentk submodule has to be bumped (the diff between those two hashes show no difference).

**Comment by mandel-macaque** [comment]
Building issues are unrelated in the PR, the checkout of the project failed on devops

**Comment by mandel-macaque** [comment]
the CI just corrected me and well as rolf, we do not need to bump the submodule, that results in the following error when building on the CI:

```
fatal: remote error: upload-pack: not our ref 6fa80b477c7e26d1c95e8c23657cd23b34d6751e
fatal: Fetched in submodule path 'external/opentk', but it did not contain 6fa80b477c7e26d1c95e8c23657cd23b34d6751e. Direct fetching of that commit failed.
```

File: src/OpenGL/OpenTK/Math/Quaternion.cs (hunk 1 above)
Comment (StephaneDelcroix) [reply src/OpenGL/OpenTK/Math/Quaternion.cs:71]: transient changes, dropping those

**Comment by rolfbjarne** [comment]
The build now fails with:

```
SceneKit/SCNMatrix4_dotnet.cs(39,25): error CS0400: The type or namespace name 'OpenTK' could not be found in the global namespace (are you missing an assembly reference?)
SceneKit/SCNMatrix4_dotnet.cs(41,25): error CS0400: The type or namespace name 'OpenTK' could not be found in the global namespace (are you missing an assembly reference?)
SceneKit/SCNMatrix4_dotnet.cs(40,26): error CS0400: The type or namespace name 'OpenTK' could not be found in the global namespace (are you missing an assembly reference?)
SceneKit/SCNMatrix4_dotnet.cs(42,28): error CS0400: The type or namespace name 'OpenTK' could not be found in the global namespace (are you missing an assembly reference?)
SceneKit/SCNMatrix4_dotnet.cs(43,29): error CS0400: The type or namespace name 'OpenTK' could not be found in the global namespace (are you missing an assembly reference?)
make[1]: *** [build/dotnet/macos/core-macos.dll] Error 1
make[1]: *** Waiting for unfinished jobs....
```

I think you changes to `SceneKit/SCNMatrix4.cs` needs to be moved to `SceneKit/SCNMatrix4_dotnet.cs`

File: src/glkit.cs (hunk 8 above)
Comment (StephaneDelcroix) [reply src/glkit.cs:254]: it's deprecated, and the keeping it require keeping Matrix[3|4]

**Comment by mandel-macaque** [comment]
Current failure is:
```
SceneKit/SCNMatrix4_dotnet.cs(832,6): error CS1061: 'Quaternion' does not contain a definition for 'ToAxisAngle' and no accessible extension method 'ToAxisAngle' accepting a first argument of type 'Quaternion' could be found (are you missing a using directive or an assembly reference?)
```

File: src/glkit.cs (hunk 8 above)
Comment (rolfbjarne) [reply src/glkit.cs:254]: Pretty much all of `GLKit` is deprecated (in favor or `MetalKit`, which is a very different API). This means that it's not a simple "don't use this class in your code, use this other class instead", it's a "port your entire rendering code to a completely different API". Also GLKit still works, so it's an entirely functional API. For these reasons I believe it's best to keep `GLKEffectPropertyTransform` (and keep/port Matrix3|4).

**Comment by rolfbjarne** [comment]
Looks like there's a syntax error in a test file:

```
/Users/builder/azdo/_work/1/s/xamarin-macios/tests/monotouch-test/SpriteKit/UniformTest.cs(66,60): error CS1513: } expected [/Users/builder/azdo/_work/1/s/xamarin-macios/tests/xharness/tmp-test-dir/monotouch-test3557/monotouch-test.csproj]
/Users/builder/azdo/_work/1/s/xamarin-macios/tests/monotouch-test/SpriteKit/UniformTest.cs(240,1): error CS1027: #endif directive expected [/Users/builder/azdo/_work/1/s/xamarin-macios/tests/xharness/tmp-test-dir/monotouch-test3557/monotouch-test.csproj]
```

**Comment by mandel-macaque** [comment]
@StephaneDelcroix the failures in the older mac OS can be ignored, yet they show that the branch is far behind main since those were fixed. Can you please rebase with main (or merge)?

**Comment by StephaneDelcroix** [comment]
> @StephaneDelcroix the failures in the older mac OS can be ignored, yet they show that the branch is far behind main since those were fixed. Can you please rebase with main (or merge)?

@mandel-macaque I rebased it before pushing (so the main is 2 days old as I type), but I'm rebasing again

**Review by dalexsoto — APPROVED** [review]

**Comment by rolfbjarne** [comment]
Test failures are unrelated (https://github.com/xamarin/maccore/issues/581).

**Review by rolfbjarne — APPROVED** [review]

**Review by rolfbjarne — APPROVED** [review]

**Pull request merged**